### DISCUSSION TITLE
For Next App Router middleware, are we supposed to just use the unseal/seal data methods?

### DISCUSSION THREAD

**Discussion by jordanpurinton** [vvo/iron-session#658] [Q&A] [answered]
I get errors when calling `save` and `destroy` methods directly in Next middleware. Are we supposed to just use the seal methods in there and modify the cookies using the request/response objects? If so, can that be better called out in the docs?

**Comment by vvo** [comment]
Hey @jordanpurinton, no it's not supposed to error. But it's true I only tested reading the session in a middleware, not writing the session. What are the errors you're getting?

**Reply by jordanpurinton** [reply]
I get errors relating to the fact that you can't write cookies using the next package unless you're in a server action or router handler

**Reply by mcl-laura** [reply]
I was calling seal/unseal myself from middleware because [getIronSession call returned empty or errors](https://github.com/vvo/iron-session/discussions/686#discussioncomment-8310654). And that ended up requiring me to install `iron-session/edge`  but after I did that, I still ended up with `bad Hmac value` error from Hapi-iron.

Are we able to use iron-session in nextJs middleware? But I am using the Pages Router... would this not work at all? Thank you! (iron-session v8.0.1

**Comment by vvo** [comment]
Hey there you can use the normal getIronSession in middleware, as seen here: https://github.com/vvo/iron-session/blob/main/examples/next/src/middleware.ts

**Comment by vvo** [comment, marked as answer]
Use `getIronSession` with the `nextProxyCookies` adapter that v9 adds, no seal/unseal by hand:

```ts
// proxy.ts (middleware.ts before Next 16)
import { NextResponse, type NextRequest } from "next/server";
import { getIronSession, nextProxyCookies } from "iron-session";

export async function proxy(request: NextRequest) {
  const response = NextResponse.next();
  const session = await getIronSession(nextProxyCookies(request, response), options);

  session.lastSeen = Date.now();
  await session.save();

  return response;
}
```

My earlier answer here was wrong. Plain `getIronSession` in middleware could read a session, but `save()` and `destroy()` silently did nothing: Next only merges a cookie into the current render when it goes through `response.cookies.set()`, and a raw `Set-Cookie` header there has no effect. The adapter does that.

v9 notes: https://github.com/vvo/iron-session/releases/tag/v9.0.0
